cmd/acraserver/listener.go

```diff
@@ -107,30 +145,149 @@ func (server *SServer) handleConnection(connection net.Conn) {
+		} else {
+			log.Infof("new connection to acraserver: <%v>", connection.RemoteAddr())
+		}
+		go func() {
+			server.cmACRA.Add(1)
+			server.handleConnection(connection)
+			server.cmACRA.Done()
+		}()
+
+	}
+}
+
+func (server *SServer) StartFromFileDescriptor(fd uintptr) {
+	var connection net.Conn
+	file := os.NewFile(fd, "/tmp/acraserver")
+	listenerFile, err := net.FileListener(file)
+	if err != nil {
+		log.WithError(err).Errorln("can't start listen for file descriptor")
+		ErrorSignalChannel <- syscall.SIGTERM
+		return
+	}
+
+	listenerWithFileDesciptor, ok := listenerFile.(network.ListenerWithFileDescriptor)
+	if !ok {
+		log.WithError(err).Errorf("File descriptor %d is not a valid socket", fd)
+		return
+	}
+	server.listenerACRA = listenerWithFileDesciptor
+	server.addListener(listenerFile)
+
 	log.Infof("start listening %s", server.config.GetAcraConnectionString())
 	for {
-		connection, err := listener.Accept()
+		connection, err = listenerWithFileDesciptor.Accept()
 		if err != nil {
-			log.WithError(err).Errorln("can't accept new connection")
-			continue
+			if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
+				log.WithError(err).Errorf("stop accepting new connections", connection)
+				return
+			}
+			log.WithError(err).Errorf("can't accept new connection (connection=%v)", connection)
 		}
 		// unix socket and value == '@'
 		if len(connection.RemoteAddr().String()) == 1 {
 			log.Infof("new connection to acraserver: <%v>", connection.LocalAddr())
 		} else {
 			log.Infof("new connection to acraserver: <%v>", connection.RemoteAddr())
 		}
-		go server.handleConnection(connection)
+		go func() {
+			server.cmACRA.Add(1)
+			server.handleConnection(connection)
+			server.cmACRA.Done()
+		}()
+	}
+}
+
+func (server *SServer) StopListeners() {
+	var (
+		err  error
+		nerr *net.TCPListener
+		ok   bool
+	)
+	if nerr, ok = server.listenerACRA.(*net.TCPListener); ok {
+		err = server.listenerACRA.(*net.TCPListener).SetDeadline(time.Now())
+		if err != nil {
+			if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
+				log.WithError(err).Errorln("unable to SetDeadLine of acra-listener")
+			}
+		}
+	}
+	if nerr != nil {
+		log.Warningf("acra-interface assigment failed - %v", nerr)
+	}
+
+	if server.listenerAPI != nil {
+		if nerr, ok = server.listenerAPI.(*net.TCPListener); ok {
+			err = server.listenerAPI.(*net.TCPListener).SetDeadline(time.Now())
+			if err != nil {
+				if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
+					log.WithError(err).Errorln("unable to SetDeadLine of api-listener")
+				}
+			}
+		} else {
+			log.Warningf("api-interface assigment failed - %v", nerr)
+		}
+	}
+}
+
+func (server *SServer) WaitConnections(duration time.Duration) {
+	log.Infof("waiting for %v connections to complete", server.ConnectionsCounter())
+	server.cmACRA.Wait()
+	if server.listenerAPI != nil {
+		server.cmAPI.Wait()
 	}
 }
 
+var WaitTimeoutError = errors.New("timeout")
```

Comment:
go has convention to name all errors with prefix `Err` at start (ErrWaitTimeout)

Reply:
done